Tariff UI: make formula field multiline capable

### Discussed in https://github.com/evcc-io/evcc/discussions/28211

<div type='discussions-op-text'>

<sup>Originally posted by **HolgerMiara** March 14, 2026</sup>
Hi everyone,

@naltatis recently added a beautiful UI to configure tariffs and forecasts. I’m trying to migrate my previous working configuration into the new UI, but I’m struggling with the single-line "formula" field.

**The Goal:**
I want to combine a dynamic spot price (Nordpool or Energienet) with time-of-use grid fees that change based on the hour and the season (Summer/Winter).

**My previous (working) multi-line logic:**
```yaml
grid:
  template: nordpool 
  region: DK2        
  currency: DKK      
  charges: 0.0130 # 0.0080 'elafgift' excl. VAT + 0.0050 surcharge on spot price excl. VAT (Vindstød)
  tax: 0.25       # 25% VAT
  # dynamic grid fees defined by local distribution grid operator;
  # below fees are taken from https://radiuselnet.dk/elnetkunder/tariffer-og-netabonnement/
  formula: |
    // Determine if it's summer (April - September)
    isSummer := ts.Month() >= 4 && ts.Month() <= 9
     // Initialize gridfee
    gridfee := 0.0000  
     // Set grid fee based on season and time of day
    if isSummer {
      if ts.Hour() < 6 {
        gridfee = 0.0976  // low load summer
      } else if ts.Hour() >= 17 && ts.Hour() < 21 {
        gridfee = 0.3808  // peak load summer
      } else {
        gridfee = 0.1465  // high load summer
      }
    } else {
      if ts.Hour() < 6 {
        gridfee = 0.0976  // low load winter
      } else if ts.Hour() >= 17 && ts.Hour() < 21 {
        gridfee = 0.8788  // peak load winter
      } else {
        gridfee = 0.2929  // high load winter
      }
    }
    (price + charges + gridfee) * (1 + tax)
``` 
**The UI Issue:**
I tried to "flatten" this into a single line for the UI formula field. However, using a ternary-style approach like condition `? true : false` results in the following error:
`cannot create tariff type 'custom': _.go:1:66: illegal character U+003F '?'`

**Questions:**

1. Is there a way to use multi-line logic (variables/if-statements) within the UI formula field?
2. If it must be a single line, what is the correct syntax for nested conditions? (It seems ? : is not supported).
3. Would it be possible to allow a "Grid Fee" template that specifically handles these common "Time-of-Use" (Tarifmodel 3.0) zones?

I am likely not alone with this; many users in Denmark (Radius/Cerius) and Germany (Paragraph 14a) face similar seasonal/hourly grid fee structures. See also: https://github.com/evcc-io/evcc/discussions/19897

Thanks for the great work on the UI!</div>

--- a/assets/js/components/Config/PropertyField.vue
+++ b/assets/js/components/Config/PropertyField.vue
@@ -75,7 +75,7 @@
 		:type="inputType"
 		:placeholder="placeholder"
 		:required="required"
-		:rows="rows || 4"
+		:rows="textareaRows"
 		:disabled="disabled"
 	/>
 	<PropertyZonesField v-else-if="zones" :id="id" v-model="value" :currency="currency" />
@@ -254,8 +254,17 @@ export default {
 			return (
 				this.rows ||
 				this.array ||
-				["accessToken", "refreshToken", "identifiers"].includes(this.property)
+				["accessToken", "refreshToken", "identifiers", "formula"].includes(this.property)
 			);
+		},
+		textareaRows() {
+			if (this.rows) return this.rows;
+			const autoGrow = this.property === "formula";
+			if (autoGrow) {
+				const lines = (this.value ?? "").split("\n").length;
+				return Math.max(1, lines);
+			}
+			return 4;
 		},
 		boolean() {
 			return this.type === "Bool";

